File: AngularJS/src/org/angular2/inspections/Angular2AnalysisHandlersFactory.java
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.inspections;

import com.intellij.codeInsight.template.Template;
import com.intellij.codeInspection.InspectionSuppressor;
import com.intellij.codeInspection.LocalQuickFix;
import com.intellij.lang.javascript.DialectDetector;
import com.intellij.lang.javascript.ecmascript6.TypeScriptAnalysisHandlersFactory;
import com.intellij.lang.javascript.psi.JSExpression;
import com.intellij.lang.javascript.psi.JSReferenceExpression;
import com.intellij.lang.javascript.psi.JSThisExpression;
import com.intellij.lang.javascript.psi.ecmal4.JSClass;
import com.intellij.lang.javascript.psi.util.JSClassUtils;
import com.intellij.lang.javascript.validation.JSReferenceChecker;
import com.intellij.lang.javascript.validation.JSReferenceInspectionProblemReporter;
import com.intellij.lang.javascript.validation.TypeScriptReferenceChecker;
import com.intellij.lang.javascript.validation.fixes.CreateJSFunctionIntentionAction;
import com.intellij.lang.javascript.validation.fixes.CreateJSVariableIntentionAction;
import com.intellij.lang.typescript.formatter.TypeScriptCodeStyleSettings;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Pair;
import com.intellij.openapi.util.Ref;
import com.intellij.psi.*;
import org.angular2.index.Angular2IndexingHandler;
import org.angular2.inspections.quickfixes.Angular2FixesFactory;
import org.angular2.lang.Angular2Bundle;
import org.angular2.lang.expr.psi.Angular2PipeReferenceExpression;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

import static com.intellij.util.ObjectUtils.notNull;

public class Angular2AnalysisHandlersFactory extends TypeScriptAnalysisHandlersFactory {

  @NotNull
  @Override
  public InspectionSuppressor getInspectionSuppressor() {
    return Angular2InspectionSuppressor.INSTANCE;
  }

  @NotNull
  @Override
  public JSReferenceChecker getReferenceChecker(@NotNull JSReferenceInspectionProblemReporter reporter) {
    return new TypeScriptReferenceChecker(reporter) {
      @Override
      protected void addCreateFromUsageFixesForCall(@NotNull JSReferenceExpression methodExpression,
                                                    boolean isNewExpression,
                                                    @NotNull ResolveResult[] resolveResults,
                                                    @NotNull List<LocalQuickFix> quickFixes) {
        if (methodExpression instanceof Angular2PipeReferenceExpression) {
          // TODO Create pipe from usage
          return;
        }
        JSExpression qualifier = methodExpression.getQualifier();
        if (qualifier == null || qualifier instanceof JSThisExpression) {
          JSClass componentClass = Angular2IndexingHandler.findComponentClass(methodExpression);
          if (componentClass != null && methodExpression.getReferenceName() != null) {
            quickFixes.add(new CreateComponentMethodIntentionAction(methodExpression));
          }
          return;
        }
        super.addCreateFromUsageFixesForCall(methodExpression, isNewExpression, resolveResults, quickFixes);
      }

      @Override
      protected Ref<String> createUnresolvedCallReferenceMessage(JSReferenceExpression methodExpression, boolean isNewExpression) {
        if (methodExpression instanceof Angular2PipeReferenceExpression) {
          return Ref.create(Angular2Bundle.message("angular.inspection.template.unresolved-pipe", methodExpression.getReferenceName()));
        }
        return super.createUnresolvedCallReferenceMessage(methodExpression, isNewExpression);
      }

      @Override
      protected void reportUnresolvedReference(@NotNull ResolveResult[] resolveResults,
                                               @NotNull JSReferenceExpression referenceExpression,
                                               @NotNull List<LocalQuickFix> quickFixes,
                                               @NotNull Ref<String> message,
                                               boolean isFunction) {
        if (referenceExpression instanceof Angular2PipeReferenceExpression) {
          Angular2FixesFactory.addUnresolvedDeclarationFixes(referenceExpression, quickFixes);
        }
        super.reportUnresolvedReference(resolveResults, referenceExpression, quickFixes, message, isFunction);
      }

      @Override
      protected boolean addCreateFromUsageFixes(JSReferenceExpression referenceExpression,
                                                ResolveResult[] resolveResults,
                                                List<LocalQuickFix> quickFixes,
                                                boolean inTypeContext,
                                                boolean ecma) {
        JSExpression qualifier = referenceExpression.getQualifier();
        if (qualifier == null || qualifier instanceof JSThisExpression) {
          JSClass componentClass = Angular2IndexingHandler.findComponentClass(referenceExpression);
          if (componentClass != null
              && referenceExpression.getReferenceName() != null) {
            quickFixes.add(new CreateComponentFieldIntentionAction(referenceExpression));
          }
          return inTypeContext;
        }
        return super.addCreateFromUsageFixes(referenceExpression, resolveResults, quickFixes, inTypeContext, ecma);
      }
    };
  }

  private static void addClassMemberModifiers(Template template, boolean staticContext, @NotNull JSClass targetClass) {
    if (DialectDetector.isTypeScript(targetClass)) {
      if (TypeScriptCodeStyleSettings.getTypeScriptSettings(targetClass).USE_PUBLIC_MODIFIER) {
        //noinspection HardCodedStringLiteral
        template.addTextSegment("public ");
      }
      if (staticContext) {
        //noinspection HardCodedStringLiteral
        template.addTextSegment("static ");
      }
    }
  }

  @NotNull
  private static SmartPsiElementPointer<JSReferenceExpression> createPointerFor(@NotNull JSReferenceExpression methodExpression) {
    return SmartPointerManager.getInstance(methodExpression.getProject()).createSmartPsiElementPointer(methodExpression);
  }

  private static class CreateComponentFieldIntentionAction extends CreateJSVariableIntentionAction {

    private final SmartPsiElementPointer<JSReferenceExpression> myRefExpressionPointer;

    CreateComponentFieldIntentionAction(JSReferenceExpression referenceExpression) {
      super(referenceExpression.getReferenceName(), true, false, false);
      myRefExpressionPointer = createPointerFor(referenceExpression);
    }

    @Override
    protected void applyFix(Project project, PsiElement psiElement, PsiFile file, Editor editor) {
      JSClass componentClass = Angular2IndexingHandler.findComponentClass(psiElement);
      assert componentClass != null;
      doApplyFix(project, componentClass, componentClass.getContainingFile(), null);
    }

    @Override
    protected JSReferenceExpression beforeStartTemplateAction(JSReferenceExpression referenceExpression,
                                                              Editor editor,
                                                              @NotNull PsiElement anchor,
                                                              boolean isStaticContext) {
      return referenceExpression;
    }

    @NotNull
    @Override
    protected Pair<JSReferenceExpression, PsiElement> calculateAnchors(PsiElement psiElement) {
      return Pair.create(myRefExpressionPointer.getElement(), psiElement.getLastChild());
    }

    @Override
    protected void addAccessModifier(Template template,
                                     @NotNull JSReferenceExpression referenceExpression,
                                     boolean staticContext,
                                     @NotNull JSClass targetClass) {
      addClassMemberModifiers(template, staticContext, targetClass);
    }
  }

  private static class CreateComponentMethodIntentionAction extends CreateJSFunctionIntentionAction {
    private final SmartPsiElementPointer<JSReferenceExpression> myRefExpressionPointer;

    CreateComponentMethodIntentionAction(JSReferenceExpression methodExpression) {
      super(methodExpression.getReferenceName(), true, false);
      myRefExpressionPointer = createPointerFor(methodExpression);
    }

    @Override
    protected void applyFix(Project project, PsiElement psiElement, PsiFile file, Editor editor) {
      JSClass componentClass = Angular2IndexingHandler.findComponentClass(psiElement);
      assert componentClass != null;
      doApplyFix(project, componentClass, componentClass.getContainingFile(), null);
    }

    @Override
    protected JSReferenceExpression beforeStartTemplateAction(JSReferenceExpression referenceExpression,
                                                              Editor editor,
                                                              @Nullable PsiElement anchor,
                                                              boolean isStaticContext) {
      return referenceExpression;
    }

    @NotNull
    @Override
    protected Pair<JSReferenceExpression, PsiElement> calculateAnchors(PsiElement psiElement) {
      return Pair.create(myRefExpressionPointer.getElement(), psiElement.getLastChild());
    }

    @Override
    protected void writeFunctionAndName(Template template,
                                        String createdMethodName,
                                        @NotNull PsiElement anchorParent,
                                        @Nullable PsiElement clazz,
                                        JSReferenceExpression referenceExpression) {
      if (referenceExpression.getQualifier() instanceof JSThisExpression) {
        createdMethodName = notNull(referenceExpression.getReferenceName(), createdMethodName);
      }
      template.addTextSegment(JSClassUtils.createClassFunctionName(createdMethodName, anchorParent));
    }

    @Override
    protected void addAccessModifier(Template template,
                                     @NotNull JSReferenceExpression referenceExpression,
                                     boolean staticContext,
                                     @NotNull JSClass targetClass) {
      addClassMemberModifiers(template, staticContext, targetClass);
    }
  }
}

File: AngularJS/src/org/angular2/inspections/Angular2AnalysisHandlersFactory.kt
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.inspections

import com.intellij.codeInsight.daemon.impl.analysis.HighlightInfoHolder
import com.intellij.codeInspection.InspectionSuppressor
import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.util.InspectionMessage
import com.intellij.lang.javascript.DialectOptionHolder
import com.intellij.lang.javascript.JSTokenTypes
import com.intellij.lang.javascript.ecmascript6.TypeScriptAnalysisHandlersFactory
import com.intellij.lang.javascript.highlighting.TypeScriptHighlighter
import com.intellij.lang.javascript.psi.JSExpression
import com.intellij.lang.javascript.psi.JSReferenceExpression
import com.intellij.lang.javascript.psi.JSThisExpression
import com.intellij.lang.javascript.psi.JSType
import com.intellij.lang.javascript.psi.JSType.TypeTextFormat.CODE
import com.intellij.lang.javascript.validation.*
import com.intellij.lang.typescript.validation.TypeScriptTypeChecker
import com.intellij.psi.PsiElement
import com.intellij.psi.ResolveResult
import com.intellij.psi.util.childLeafs
import com.intellij.psi.util.elementType
import com.intellij.psi.util.parentOfType
import com.intellij.util.ProcessingContext
import org.angular2.codeInsight.Angular2HighlightingUtils.TextAttributesKind.NG_PIPE
import org.angular2.codeInsight.Angular2HighlightingUtils.withColor
import org.angular2.codeInsight.blocks.BLOCK_FOR
import org.angular2.entities.Angular2ComponentLocator
import org.angular2.inspections.quickfixes.*
import org.angular2.lang.Angular2Bundle
import org.angular2.lang.expr.psi.*
import org.angular2.lang.html.psi.Angular2HtmlPropertyBinding
import org.angular2.signals.Angular2SignalUtils

class Angular2AnalysisHandlersFactory : TypeScriptAnalysisHandlersFactory() {

  override fun getInspectionSuppressor(): InspectionSuppressor {
    return Angular2InspectionSuppressor
  }

  override fun <T : Any?> getTypeChecker(problemReporter: JSProblemReporter<T>): JSTypeChecker =
    object : TypeScriptTypeChecker(problemReporter) {

      override fun getFixes(expr: JSExpression?,
                            declaredJSType: JSType,
                            elementToChangeTypeOf: PsiElement?,
                            expressionJSType: JSType,
                            context: ProcessingContext?,
                            holder: DialectOptionHolder?): Collection<LocalQuickFix> {
        val quickFixes = super.getFixes(expr, declaredJSType, elementToChangeTypeOf, expressionJSType, context, holder)
        if (elementToChangeTypeOf is Angular2HtmlPropertyBinding) {
          return Angular2FixesFactory.getCreateInputTransformFixes(elementToChangeTypeOf,
                                                                   expressionJSType.substitute().getTypeText(CODE)) + quickFixes
        }
        return quickFixes
      }
    }

  override fun getReferenceChecker(reporter: JSProblemReporter<*>): JSReferenceChecker =
    object : TypeScriptReferenceChecker(reporter) {
      override fun addCreateFromUsageFixesForCall(methodExpression: JSReferenceExpression,
                                                  isNewExpression: Boolean,
                                                  resolveResults: Array<ResolveResult>,
                                                  quickFixes: MutableList<in LocalQuickFix>) {
        if (methodExpression is Angular2PipeReferenceExpression) {
          // TODO Create pipe from usage
          return
        }
        val qualifier = methodExpression.qualifier
        if (qualifier == null || qualifier is JSThisExpression) {
          val componentClass = Angular2ComponentLocator.findComponentClass(methodExpression)
          if (componentClass != null && methodExpression.referenceName != null) {
            quickFixes.add(CreateComponentMethodIntentionAction(methodExpression))
            if (Angular2SignalUtils.supportsSignals(componentClass)) {
              quickFixes.add(CreateComponentSignalIntentionAction(methodExpression))
            }
          }
          return
        }
        super.addCreateFromUsageFixesForCall(methodExpression, isNewExpression, resolveResults, quickFixes)
      }

      @InspectionMessage
      override fun createUnresolvedCallReferenceMessage(methodExpression: JSReferenceExpression, isNewExpression: Boolean): String {
        return if (methodExpression is Angular2PipeReferenceExpression) {
          Angular2Bundle.htmlMessage(
            "angular.inspection.unresolved-pipe.message",
            methodExpression.getReferenceName()!!.withColor(NG_PIPE, methodExpression)
          )
        }
        else super.createUnresolvedCallReferenceMessage(methodExpression, isNewExpression)
      }

      override fun reportUnresolvedReference(resolveResults: Array<ResolveResult>,
                                             referenceExpression: JSReferenceExpression,
                                             quickFixes: MutableList<LocalQuickFix>,
                                             @InspectionMessage message: String,
                                             isFunction: Boolean,
                                             inTypeContext: Boolean) {
        super.reportUnresolvedReference(resolveResults, referenceExpression, quickFixes, message, isFunction, inTypeContext)
      }

      override fun addCreateFromUsageFixes(referenceExpression: JSReferenceExpression,
                                           resolveResults: Array<ResolveResult>,
                                           quickFixes: MutableList<in LocalQuickFix>,
                                           inTypeContext: Boolean,
                                           ecma: Boolean): Boolean {
        val qualifier = referenceExpression.qualifier
        if (qualifier == null || qualifier is JSThisExpression) {
          val componentClass = Angular2ComponentLocator.findComponentClass(referenceExpression)
          if (componentClass != null && referenceExpression.referenceName != null) {
            quickFixes.add(CreateComponentFieldIntentionAction(referenceExpression))
            if (referenceExpression.parentOfType<Angular2EmbeddedExpression>() is Angular2Action) {
              quickFixes.add(CreateDirectiveOutputIntentionAction(referenceExpression, referenceExpression.referenceName!!))
            }
          }
          return inTypeContext
        }
        return super.addCreateFromUsageFixes(referenceExpression, resolveResults, quickFixes, inTypeContext, ecma)
      }

    }

  override fun createKeywordHighlighterVisitor(holder: HighlightInfoHolder,
                                               dialectOptionHolder: DialectOptionHolder): JSKeywordHighlighterVisitor =
    object : TypeScriptKeywordHighlighterVisitor(holder) {
      override fun visitElement(element: PsiElement) {
        when (element) {
          is Angular2BlockParameter -> if (element.block?.getName() == BLOCK_FOR && element.isPrimaryExpression)
            element.node.findChildByType(JSTokenTypes.IDENTIFIER)
              ?.let { highlightKeyword(it, TypeScriptHighlighter.TS_KEYWORD) }
              ?.let { myHolder.add(it) }
          is Angular2DeferredTimeLiteralExpression -> element.childLeafs
            .find { it.elementType == JSTokenTypes.IDENTIFIER }
            ?.takeIf { it.text == "s" || it.text == "ms" }
            ?.let { highlightKeyword(it.node, TypeScriptHighlighter.TS_NUMBER) }
            ?.let { myHolder.add(it) }
          else -> super.visitElement(element)
        }
      }
    }

}
